Added commit termination callback functionality in BackoffComitter

accelerated opened the pull request:
Added commit termination callback functionality in the `BackoffComitter` class similar to `BufferedProducer`.

mfontanini commented:
I'm not sure about this one. You can currently implement this by tracking the failures in the error callback: the idea of these classes is you can implement your own logic outside of them via the callback. It seems like you're trying to make these classes fit what you want rather than making what you want fit into what these classes can do. You could implement your own "SmartBackoffCommitter" that tracks failures internally via the callback and indicates what happened after every commit attempt instead.

accelerated commented:
Couple of thougths:
- my initial thinking was to break ABI and return an error code directly. The commit member function is synchronous and there's really no reason to return the error via a callback like the one i proposed. So if you are ok with that we can change the signature to return int or bool. Much simpler!
- the error callback does not provide any partition info, which makes tracking impossible. If I commit on different partitions from diff threads, your subclass idea won't work. Also my library is generic enough that i don't want to force people to register an offset commit callback just to make this work. Not to mention now i have to correlate error callbacks with offset commit callback from rdkafka which has partition info.
- maybe not a super valid reason but it's in the same lines as the flush termination and delivery termination callbacks in the buffered producer.

mfontanini commented:
Wouldn't something like just throwing if you reach the max retries in `BackoffPerformed::perform` fix your problem? It feels like "This callback will be executed when all commit attempts have been exhausted and the commit still failed, signalling the end of all retries" (taken from the callback's docs) is exactly what exceptions are for: something bad happened and you can't handle it locally. Is there something your solution is achieving that an exception wouldn't? I may be missing something here.

accelerated commented:
I can definitely throw if that's what you prefer. However there are 2 places for throwing, when max retries have ended and when the error callback returns false. Lmk if you are ok with this.

accelerated commented:
Changes wrt to the first implementation:
- No API changes and no termination callbacks.
- `BackoffPerformer::perform` throws `ConsumerException`.
- Added `BackoffPerformer::ReturnType` to handle all callback possible continuations.
- Default constructor for `Error` type. 
- Added `BackoffCommitter::commit()` to handle all assigned partitions. Calls `Consumer::commit()` under the hood.

accelerated commented on include/cppkafka/utils/backoff_performer.h:
```diff
@@ -134,6 +148,8 @@ class CPPKAFKA_API BackoffPerformer {
             backoff = increase_backoff(backoff);
         }
+        // No more retries left or we have a terminal error.
+        throw ConsumerException(rt.error_);
```
Ok will add variadic templates and change the last exception to `Exception` which only takes a string. However I don't see how we can circumvent not using ReturnType since we still need to detect when the functor ends because it was successful or because the error callback returned false. Perhaps an int like before with `0=success+abort, 1=fail+abort,-1=fail+continue`.

mfontanini replied:
Yeah, you're right. I'd rather add an enum for this though, using `int`s is not ideal as you have no clue what those mean out of context.

accelerated replied:
Ok before you responded I made this other change which doesn't need any enums. Take a look please. If you still want me to use enums I will. But the last commit requires the least API changes.

mfontanini commented on include/cppkafka/utils/backoff_committer.h:
```diff
@@ -127,25 +135,30 @@ class CPPKAFKA_API BackoffCommitter : public BackoffPerformer {
-            if (ex.get_error() == RD_KAFKA_RESP_ERR__NO_OFFSET) {
-                return true;
+            if (error == RD_KAFKA_RESP_ERR__NO_OFFSET) {
+                throw ex; //abort
```
I think this should probably just return `true`? At least that's what was happening before. I don't really like this error, as it's a non-error error. Committing when there's no offsets is a no-op so I don't really know why it should fail. In my mind this error should be ignored everywhere. Can you think of any case where you'd actually want to be notified that there's nothing to commit?

accelerated replied:
I see your point. Not sure really. The rdkafka documentation states:
 >If no partitions had valid offsets to commit this callback will be called with \p err == RD_KAFKA_RESP_ERR__NO_OFFSET which is not to be considered an error.

So yes I think it's safe to just return true in this case.

mfontanini replied:
lol nice. Yeah, I don't really know why this returns a non-error error. It should IMO just return success as the commit itself didn't fail. This basically means every place where you commit you have to check for this special error, which makes it really annoying for users.

But going back to this, yeah, this should probably keep returning true like before.

accelerated commented:
Seems like the rdkafka version you are trying to wget is wrong...all tests get stuck on this. Also your min, max version numbers need updating int he CMakeLists.txt. I would be good to add a patch version as well since you are now at 0.3.1.

accelerated commented:
BTW would you consider throwing `HandleException(RD_KAFKA_RESP_ERR__FAIL)` from the BackoffPerformer::perform() instead of the more generic `Exception(msg)`?

mfontanini commented:
I noticed the kafka download issue, it looks like whatever mirror that's using is not longer working.

Regarding the exception, I feel like this should probably have its own exception type (e.g. `MessageCommitException`). Using `HandleException` with some "made up" error code doesn't make much sense to me.

accelerated commented:
I renamed it `ActionTerminatedException` since the `BackoffPerformer` is a generic class, not intended for only committing.

accelerated commented:
@mfontanini would it be possible to merge this? Anything else that needs done?

mfontanini commented:
Sorry for the delay and thanks!
